### PR TITLE
Implement `GoodJob::Batch`

### PR DESCRIPTION
**Pull request by bensheldon**
Connects to #417, #691.

- [x] Add description field
- [x] Add Dashboard view
- [x] Add cleanup for batches
- [x] Rename batch `params` to `properties`
- [x] Make migration-less patch-release compatible

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -396,6 +397,74 @@ end
...
+    # When these jobs have finished, it will enqueue your `MyBatchCallbackJob.perform_later(batch, options)`
+    class MyBatchCallbackJob < ApplicationJob
+      def perform(batch, options = {})
+        # Do something with the <GoodJob::Batch> instance; options is a hash is reserved for future use
+        batch.properties[:user] # => <User id: 1, ...>
+      end
+    end
+    ```
+
+- Jobs can be added to an existing batch. Jobs in a batch are enqueued and performed immediately, though the final callback job will not be enqueued until `GoodJob::Batch#enqueue` is called.
```
Comment (mollerhoj) [inline README.md:421]: I find these instructions unclear. To me, it looks like the code below contains a race condition. What happens if all the jobs finishes before `enqueue` is called?.

On another note: What happens if you call `enqueue` twice on the same batch?
What happens if you call `enqueue` on a batch without any jobs?

File: README.md [hunk 2]
```diff
@@ -396,6 +397,74 @@ end
...
+    batch = GoodJob::Batch.add do
+      10.times { MyJob.perform_later }
+    end
+    batch.add do
+      10.times { OtherJob.perform_later }
+    end
+    batch.enqueue(MyBatchCallbackJob, age: 42)
+    ```
+
+- If you need to access the batch within a model, include [`GoodJob::ActiveJobExtensions::Batches`](lib/good_job/active_job_extensions/batches.rb) in your job class:
```
Comment (mollerhoj) [inline README.md:433]: Am I allowed to add jobs dynamically to the batch? (If yes, let's add an example in the documentation, this is extremely useful.)

**Review by mollerhoj — COMMENTED** [review]
I have not have time yet to go over the code in detail, but I've written some questions that popped up when reading the documentation.

All in all, this looks very, very good to me. Nice API, seems like a pretty clean implementation, and an absolute KILLER feature.

I think the documentation would be even better with an example of a complex workflow (using batches to run both serial and parallel jobs in succession.) E.g, show how one would implement this workflow:


workflow

File: README.md (hunk 1 above)
Comment (bensheldon) [reply README.md:421]: Great callout! I will cover that in the documentation. Briefly:

- the callback job won't be enabled until `#enqueue` is called on the batch. 
- calling `#enqueue` on a batch without any jobs will immediately enqueue the callback job
- `#enqueue` can be called multiple times, which will re-enable the callback job to be enqueued
- `#add` will add jobs to the batch without triggering an enqueue

Do you think this kind of example makes sense for creating a multistage batch?

```ruby
class MyBatchJob < ApplicationJob
  def perform(batch, options = {})
    if batch.properties[:stage] == 1
      batch.properties[:stage] = 2
      batch.enqueue do
        10.times { OtherJob.perform_later}
      end
    elsif batch.properties[:stage] == 2
      batch.properties[:stage] = 3
      batch.enqueue do
        OtherJob.perform_later
      end
    else
      # all done
    end 
  end
end

GoodJob::Batch.enqueue(MyBatchJob, stage: 1)
```

**Pull request merged**